rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,654 @@
+		ResourceKind:  types.KindNode,
+		BackendPrefix: backend.Key(nodesPrefix, namespace),
+		MarshalFunc:   services.MarshalServer,
+		UnmarshalFunc: services.UnmarshalServer,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newSvc, err := generic.NewService(&generic.ServiceConfig[types.Server]{
+		Backend:       b,
+		ResourceKind:  types.KindNode,
+		BackendPrefix: backend.Key(nodesPrefixV2, namespace),
+		MarshalFunc:   func(t types.Server) ([]byte, error) {
+			return proto.Marshal(t)
+		}
+		UnmarshalFunc: func(types.Server) ([]byte, error) { return nil, nil }
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	servers, err := oldSvc.GetResources(ctx)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	for _, s := range servers {
+		if err := newSvc.CreateResource(ctx, s); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (e exampleMigration) Down(ctx context.Context, b backend.Backend) error {
+    oldSvc, err := generic.NewService(&generic.ServiceConfig[types.Server]{
+    Backend:       b,
+	ResourceKind:  types.KindNode,BackendPrefix: backend.Key(nodesPrefix, namespace),
+    MarshalFunc:   services.MarshalServer,
+    UnmarshalFunc: services.UnmarshalServer,
+    })
+    if err != nil {
+        return trace.Wrap(err)
+    }
+
+    newSvc, err := generic.NewService(&generic.ServiceConfig[types.Server]{
+    Backend:       b,
+    ResourceKind:  types.KindNode,
+    BackendPrefix: backend.Key(nodesPrefixV2, namespace),;''
+    MarshalFunc:   func(t types.Server) ([]byte, error) {
+        return proto.Marshal(t)
+    }
+    UnmarshalFunc: func(b []byte) (types.Server, error) {
+        var s types.ServerV2
+        if err :=  proto.Unmarshal(b, s); err != nil {
+            return nil, err,
+        }
+        return s, nil
+    })
+    if err != nil {
+        return trace.Wrap(err)
+    }
+
+    servers, err := newSvc.GetResources(ctx)
+    if err != nil {
+        if trace.IsNotFound(err) {
+            return nil
+        }
+        return trace.Wrap(err)
+    }
+
+    for _, s := range servers {
+        if err := oldSvc.CreateResource(ctx, s); err != nil {
+            return trace.Wrap(err)
+        }
+	}
+
+    return nil
+}
+```
+
+</details>
+
+Migration registration requires the version of the migration and the implementation, if a migration with the same
+version is already exists then registration will fail.
+
+```go
+    if err := migrate.Register(1, exampleMigration{}); err != nil {
+		return err
+    }
+```
+
+##### Backend Services
+
+The generic backend service will be extended to handle support for optimistic locking, performing version checking, and
+downgrading resources. This will reduce the burden on each custom backend service and ease the developer experience. To
+opt in to the new behavior a resources backend service just needs to ensure that it is using the
+`lib/services/generic.Service`.
+
+### Option 2: Phased Migrations
```

Comment:
This option makes most sense to me. It is much simpler than option 1 and scopes it to cloud, where we control the upgrade process.